File: backend/mno-integration/adapters/fallback.py
import os
import csv
import logging
from datetime import datetime

logger = logging.getLogger(__name__)

class SFTPFallbackManager:
    """
    Manages writing offline deductions to a CSV cache.
    A separate cron job will use Paramiko to upload these to the MNO's SFTP server
    when the REST API outage exceeds 5 minutes.
    """
    
    def __init__(self, carrier_name: str, cache_dir: str = "/tmp/smoothswitch/sftp"):
        self.carrier_name = carrier_name
        self.cache_dir = cache_dir
        os.makedirs(self.cache_dir, exist_ok=True)
        
        # Format: MTN_fallback_20231024.csv
        self.current_file = os.path.join(
            self.cache_dir, 
            f"{self.carrier_name}_fallback_{datetime.now().strftime('%Y%m%d')}.csv"
        )
        self._ensure_header()

    def _ensure_header(self):
        if not os.path.exists(self.current_file):
            with open(self.current_file, mode='w', newline='') as f:
                writer = csv.writer(f)
                writer.writerow(["timestamp", "imsi", "deduction_mb", "status"])

    def append_record(self, imsi: str, amount_mb: float) -> bool:
        try:
            with open(self.current_file, mode='a', newline='') as f:
                writer = csv.writer(f)
                writer.writerow([
                    datetime.utcnow().isoformat(),
                    imsi,
                    amount_mb,
                    "PENDING_SFTP"
                ])
            logger.info(f"[{self.carrier_name}] Fallback record written for IMSI {imsi}")
            return True
        except Exception as e:
            logger.error(f"[{self.carrier_name}] Failed to write SFTP fallback: {e}")
            return False

File: backend/mno-integration/adapters/fallback.go
package adapters

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"
)

type SFTPFallbackManager struct {
	carrierName string
	cacheDir    string
	mu          sync.Mutex
}

func NewSFTPFallbackManager(carrierName string, cacheDir string) *SFTPFallbackManager {
	if cacheDir == "" {
		cacheDir = "/tmp/smoothswitch/sftp"
	}
	os.MkdirAll(cacheDir, 0755)

	return &SFTPFallbackManager{
		carrierName: carrierName,
		cacheDir:    cacheDir,
	}
}

func (m *SFTPFallbackManager) currentFile() string {
	dateStr := time.Now().Format("20060102")
	return filepath.Join(m.cacheDir, fmt.Sprintf("%s_fallback_%s.csv", m.carrierName, dateStr))
}

func (m *SFTPFallbackManager) ensureHeader(filename string) error {
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		f, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			return err
		}
		defer f.Close()

		writer := csv.NewWriter(f)
		defer writer.Flush()
		return writer.Write([]string{"timestamp", "imsi", "deduction_mb", "status"})
	}
	return nil
}

func (m *SFTPFallbackManager) AppendRecord(imsi string, amountMB float64) bool {
	m.mu.Lock()
	defer m.mu.Unlock()

	filename := m.currentFile()
	if err := m.ensureHeader(filename); err != nil {
		log.Printf("[%s] Failed to create SFTP fallback header: %v", m.carrierName, err)
		return false
	}

	f, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		log.Printf("[%s] Failed to open SFTP fallback file: %v", m.carrierName, err)
		return false
	}
	defer f.Close()

	writer := csv.NewWriter(f)
	defer writer.Flush()

	record := []string{
		time.Now().UTC().Format(time.RFC3339),
		imsi,
		fmt.Sprintf("%f", amountMB),
		"PENDING_SFTP",
	}

	if err := writer.Write(record); err != nil {
		log.Printf("[%s] Failed to write SFTP fallback: %v", m.carrierName, err)
		return false
	}

	log.Printf("[%s] Fallback record written for IMSI %s", m.carrierName, imsi)
	return true
}
